### PR TITLE
Stateful failover

### LINKED ISSUE #148: Separate voter role that will help to do failover
In GitLab by @racktear on Jul 5, 2018, 20:00

There should be a separate voter role that works like this:

- If master or replica sees that their peer has died (through membership), it asks the voter who should be leader
- The voter then decides and replies with the decision based on what it sees
- If master or replica looses connection to the voter, it continues to operate as before
- If master or replica looses connection to the voter and their peer, it becomes read-only until the voter reappears
- If cluster configuration has no voter roles, the order of master-slave is decided by lexicographically comparing live node uuids in the replicaset
- All nodes periodically poll the voter for role changes

### PR DESCRIPTION
**Pull request by rosik**
This patch introduces new failover mode - stateful.

There are 3 main concepts:

**Internal coordinator** is a new role, which makes decisions regarding leadership. Earlier it was a part of every instances' failover module, but now it's split away. There may be only one active coordinator in cluster at a time. Its uniqueness is ensured by external storage which manages the lock and saves appointments.

**External storage** (`kingdom.lua`) is a stand-alone Tarantool instance which provides locking mechanism and keeps decisions made by the coordinator.

**Failover module** (the old one) operates on every instance in cluster and gathers leadership information for others modules. It was refactored too, and now it can be described with 4 functions with clearly separated responsibilities:

- `_get_appointments_*` generates leadership map by itself or polls it from external storage depending on the mode setting (disabled/eventual/stateful).
- `accept_appointments()` just refreshes the cache and tracks if anything changed.
- `failover_loop` (a fiber) repeatedly gets new appointments and accepts them using corresponding functions from the above.
- `cfg` is called from `confapplier.apply_config()` (on restart or on committing new clusterwide configuration). At first it gets appointments synchronously and then starts the failover loop.

I didn't forget about

- [x] Tests
- [x] Changelog
- [x] Documentation

Close #148

### CODE REVIEW

File: cartridge/failover-slave.lua [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+--
+-- * In `disabled` mode the leader is the first server configured in
+--   `topology.replicsets[].master` array.
+-- * The `eventual` mode has been deprecated since v2.0.1-69 and now it
+--   is equivalent for `disabled`.
+-- * When `enabled`, failover polls leaders map from the coordinator.
+--
+-- This module behavior depends on the instance state.
+--
+-- From the very beginig it reports `is_rw() == false`,
```
Comment (artur-barsegyan) [inline cartridge/failover-slave.lua]: typo

File: cartridge/failover.old.lua [hunk 2]
```diff
@@ -0,0 +1,267 @@
+error([[This module is preserved for reference purpose only,
```
Comment (olegrok) [inline cartridge/failover.old.lua]: Why don't you want to copy it to e.g. wiki page? I don't like approach when you preserve unused code. You have a git and could to move to any point into a history "for reference purpose"

File: kingdom.lua [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+    local audit = box.space.leader_audit
+    local latest_audit = audit:select(nil, {iterator = 'LT', limit = 1})[1]
+    local latest_ordinal = latest_audit and latest_audit.ordinal or 0
+    local session = box.session.storage
+
+    if session.ordinal == nil then
+        session.ordinal = latest_ordinal
+        return _G.get_leaders()
+    elseif session.ordinal > latest_ordinal then
+        error('Impossiburu! (session_ordinal > latest_ordinal)')
```
Comment (olegrok) [inline kingdom.lua]: `Impossible`?

File: kingdom.lua [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+function _G.get_lock_delay()
+    return LOCK_DELAY
+end
+
+box.cfg({
+    work_dir = work_dir,
+    listen = os.getenv('TARANTOOL_LISTEN'),
+})
+
+box.schema.user.grant('guest', 'super', nil, nil, {if_not_exists = true})
```
Comment (olegrok) [inline kingdom.lua]: You have a limited number of functions/spaces here. Let's grant explicitly access to each space/function.

Do I assume correctly that everybody could connect to such instance and "eval" "os.execute"?

File: kingdom.lua [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+end
+
+function _G.longpoll(timeout)
+    checks('?number')
+    if timeout == nil then
+        timeout = 1
+    end
+
+    local audit = box.space.leader_audit
+    local latest_audit = audit:select(nil, {iterator = 'LT', limit = 1})[1]
```
Comment (olegrok) [inline kingdom.lua]: `audit:min()`?

File: kingdom.lua [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+    end
+    return ret
+end
+
+function _G.get_king()
+    if lock.king ~= nil then
+        return lock.king
+    end
+
+    local last_king = box.space.king_audit:select(nil,
```
Comment (olegrok) [inline kingdom.lua]: `audit:min()`?

File: kingdom.lua [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+    )[1]
+    if last_king ~= nil then
+        return {
+            uuid = last_king.uuid,
+            uri = last_king.uri,
+        }
+    end
+end
+
+-- require('console').start()
```
Comment (olegrok) [inline kingdom.lua]: Drop it if unused

File: kingdom.lua (hunk 3 above)
Comment (rosik) [reply kingdom.lua]: https://knowyourmeme.com/memes/impossibru

File: test/integration/kingdom_test.lua [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+    t.assert_error_msg_equals(
+        "Duplicate key exists in unique index 'ordinal'" ..
+        " in space 'leader_audit'",
+        c.call, c, 'set_leaders', {{{'A', 'a2'}, {'A', 'a3'}}}
+    )
+end
+
+function g.test_longpolling()
+    local c1 = connect(g.kingdom   )
```
Comment (olegrok) [inline test/integration/kingdom_test.lua]: `connect(g.kingdom   )` -> `connect(g.kingdom)`

File: cartridge/roles/coordinator.lua [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+
+    log.info('Lock released')
+    return true
+end
+
+local function connect_loop(uri)
+    checks('string')
+    repeat
+        local t1 = fiber.time()
+        local ok, err = errors.pcall('E', connect, uri)
```
Comment (olegrok) [inline cartridge/roles/coordinator.lua]: Is it expected "E" type?

File: cartridge/roles/coordinator.lua [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+vars:new('connect_fiber', nil)
+vars:new('topology_cfg', nil)
+vars:new('conn', nil)
+vars:new('options', {
+    RECONNECT_PERIOD = 5,
+    IMMUNITY_TIMEOUT = 15,
+    NETBOX_CALL_TIMEOUT = 1,
+})
+
+vars:new('healthcheck', function(members, instance_uuid)
```
Comment (olegrok) [inline cartridge/roles/coordinator.lua]: Why not `local healthcheck `? I don't see any mentions in test (it's a single reason I think to store it into vars)

Comment (rosik) [reply cartridge/roles/coordinator.lua]: It's an extension point. I fear someone would need to monkeypatch it and use another predicate.

Comment (olegrok) [reply cartridge/roles/coordinator.lua]: Ok. Sometimes it's better to write some words in comment to eliminate possible misunderstanding. Thanks for explanation.

File: test/integration/kingdom_test.lua [hunk 11]
```diff
@@ -0,0 +1,148 @@
+local fio = require('fio')
+local uuid = require('uuid')
+-- local fiber = require('fiber')
```
Comment (Kasen) [inline test/integration/kingdom_test.lua]: Comment

File: cartridge/failover.lua [hunk 12]
```diff
@@ -75,43 +98,73 @@ local function _get_health_map(topology_cfg, mode)
                 member.payload.state == 'ConfiguringRoles' or
                 member.payload.state == 'RolesConfigured'
```
[the remark this answers is not in the document]
Comment (rosik) [reply cartridge/failover.lua]: Discussed orally, resolved

File: kingdom.lua [hunk 13]
```diff
@@ -0,0 +1,223 @@
...
+        box.space.leader_audit:insert({
+            ordinal, fiber.time(), replicaset_uuid, instance_uuid
+        })
+        log.info('New leader %s -> %s', replicaset_uuid, instance_uuid)
+    end
+    notification:broadcast()
+    return true
+end
+function _G.set_leaders(...)
+    return box.atomic(set_leaders, ...)
```
[the remark this answers is not in the document]
Comment (rosik) [reply kingdom.lua]: Fixed - extracted them into separate fiber

File: cartridge/failover.lua [hunk 14]
```diff
@@ -75,43 +98,73 @@ local function _get_health_map(topology_cfg, mode)
...
-                ret.potential_leaders[instance_uuid] = false
+                appointments[replicaset_uuid] = instance_uuid
+                break
             end
         end
 
-        if ret.active_leaders[replicaset_uuid] == nil then
-            ret.active_leaders[replicaset_uuid] = leaders[1]
+        if appointments[replicaset_uuid] == nil then
+            appointments[replicaset_uuid] = leaders[1]
```
[the remark this answers is not in the document]
Comment (rosik) [reply cartridge/failover.lua]: Discussed orally. It's ok in eventual mode

File: cartridge/failover.lua [hunk 15]
```diff
@@ -75,43 +98,73 @@ local function _get_health_map(topology_cfg, mode)
...
+end
+
+--- Get appointments from external storage.
+-- Used in 'stateful' failover mode.
+-- @function _get_appointments_stateful_mode
+-- @local
+local function _get_appointments_stateful_mode(conn, timeout)
+    checks('table', 'number')
+    local appointments, err = errors.netbox_call(
+        conn, 'longpoll', {timeout*0.8}, {timeout = timeout}
```
[the remark this answers is not in the document]
Comment (rosik) [reply cartridge/failover.lua]: Refactored

File: cartridge/failover.lua [hunk 16]
```diff
@@ -141,36 +194,43 @@ local function apply_config(mod)
...
 
-        -- The event may arrive during two-pahse commit is in progress.
+        if not accept_appointments(appointments) then
+            -- nothing changed
+            goto start_over
+        end
+
+        -- The event may arrive during two-phase commit is in progress.
         -- We should wait for the appropriate state.
         local state = confapplier.wish_state('RolesConfigured', math.huge)
```
[the remark this answers is not in the document]
Comment (rosik) [reply cartridge/failover.lua]: Yes, but no. It's infinite only if `ConfiguringRoles` state stuck forever (which is unlikely). Whenever it changes (either to `RolesConfigured` or to `OperationError` state) the function returns.

File: cartridge/failover.lua [hunk 17]
```diff
@@ -141,36 +194,43 @@ local function apply_config(mod)
...
             log.info('Skipping failover step - state is %s', state)
             goto start_over
         end
 
+        log.info('Failover triggered')
         confapplier.set_state('ConfiguringRoles')
-        local ok, err = FailoverError:pcall(function()
-            log.info('Failover triggered')
 
+        local ok, err = FailoverError:pcall(function()
```
[the remark this answers is not in the document]
Comment (rosik) [reply cartridge/failover.lua]: We will in #116

File: cartridge/topology.lua [hunk 18]
```diff
@@ -151,36 +151,36 @@ local function validate_schema(field, topology)
...
-                type(topology.failover.coordinator_uri) == 'string',
-                '%s.failover.coordinator_uri must be a string, got %s',
-                field, type(topology.failover.coordinator_uri)
+                type(topology.failover.storage_uri) == 'string',
+                '%s.failover.storage_uri must be a string, got %s',
+                field, type(topology.failover.storage_uri)
             )
 
-            local parts = uri.parse(topology.failover.coordinator_uri)
+            local parts = uri.parse(topology.failover.storage_uri)
```
Comment (HustonMmmavr) [inline cartridge/topology.lua]: Maybe use pool.format_uri, something like that:
```
local _, err = pool.format.uri(uri)
  e_config:assert(
                not err,
                '%s.failover.coordinator_uri: ' .. err.err,
                ...
)
```

**Pull request merged**